This page gives access to the data of the study. The whole data set is available in two forms:
<ol>
  <li>for <b>download</b> and offline analysis. There are three ways to query the data set:
    <ol>
      <li><b>Clinical and summary data:</b> Tables containing clinical information (viral load, CD4+ counts, HLA types, estimates of time of infection), the patient reference sequence for mapping, and the single nucleotide variant frequency tables.</li>
      <li><b>Haplotypes:</b> Extended genetic variants, including minor ones, covering standard or custom genomic regions.</li>
      <li><b>Sequencing reads:</b> The filtered sequencing reads, mapped against the single fragments of the patient-specific reference. Download the genbank (GB) version of the reference to find PCR fragment annotations.</li>
     </ol>
  </li>
  <li>for <b>online</b> analysis or embedding into another website. The recommended interface in this case is the <b>RESTful API</b>.</li>
<ol>
TODO: explain the API!
